File: backend/app/api/v1/agents.py
from uuid import UUID

from fastapi import APIRouter, HTTPException
from fastapi.responses import StreamingResponse

from app.models.agent import (
    Agent,
    AgentCreate,
    AgentExecuteInput,
    AgentExecuteOutput,
    AgentUpdate,
)
from app.services.agent_service import get_agent_service

router = APIRouter(prefix="/agents", tags=["agents"])


@router.get("", response_model=list[Agent])
async def get_agents(skip: int = 0, limit: int = 20):
    """에이전트 목록 조회"""
    service = get_agent_service()
    return await service.list_agents(skip=skip, limit=limit)


@router.get("/{agent_id}", response_model=Agent)
async def get_agent(agent_id: UUID):
    """에이전트 단일 조회"""
    service = get_agent_service()
    agent = await service.get_agent(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.post("", response_model=Agent, status_code=201)
async def create_agent(data: AgentCreate):
    """에이전트 생성"""
    service = get_agent_service()
    return await service.create_agent(data)


@router.patch("/{agent_id}", response_model=Agent)
async def update_agent(agent_id: UUID, data: AgentUpdate):
    """에이전트 수정"""
    service = get_agent_service()
    agent = await service.update_agent(agent_id, data)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return agent


@router.delete("/{agent_id}", status_code=204)
async def delete_agent(agent_id: UUID):
    """에이전트 삭제"""
    service = get_agent_service()
    deleted = await service.delete_agent(agent_id)
    if not deleted:
        raise HTTPException(status_code=404, detail="Agent not found")


@router.post("/{agent_id}/execute", response_model=AgentExecuteOutput)
async def execute_agent(agent_id: UUID, input_data: AgentExecuteInput):
    """에이전트 실행 (비스트리밍)"""
    service = get_agent_service()
    try:
        return await service.execute_agent(agent_id, input_data)
    except ValueError as e:
        raise HTTPException(status_code=404, detail=str(e))


@router.get("/{agent_id}/executions")
async def get_executions(agent_id: UUID, skip: int = 0, limit: int = 20):
    """에이전트 실행 이력 조회"""
    service = get_agent_service()
    agent = await service.get_agent(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")
    return await service.get_executions(agent_id, skip=skip, limit=limit)


@router.post("/{agent_id}/execute/stream")
async def execute_agent_stream(agent_id: UUID, input_data: AgentExecuteInput):
    """에이전트 실행 (스트리밍 SSE)"""
    service = get_agent_service()
    agent = await service.get_agent(agent_id)
    if not agent:
        raise HTTPException(status_code=404, detail="Agent not found")

    async def event_generator():
        try:
            async for chunk in service.execute_agent_stream(agent_id, input_data):
                yield f"data: {chunk}\n\n"
            yield "data: [DONE]\n\n"
        except Exception as e:
            yield f"data: [ERROR] {str(e)}\n\n"

    return StreamingResponse(
        event_generator(),
        media_type="text/event-stream",
        headers={
            "Cache-Control": "no-cache",
            "Connection": "keep-alive",
        },
    )

File: frontend/src/lib/api/agents.ts
import { apiClient } from './client'
import type {
  Agent,
  AgentCreate,
  AgentUpdate,
  AgentExecuteInput,
  AgentExecuteOutput,
} from '@/types/agent'

const API_URL = process.env.NEXT_PUBLIC_API_URL || 'http://localhost:8000'

export async function getAgents(skip = 0, limit = 20): Promise<Agent[]> {
  return apiClient<Agent[]>(`/api/v1/agents?skip=${skip}&limit=${limit}`)
}

export async function getAgent(id: string): Promise<Agent> {
  return apiClient<Agent>(`/api/v1/agents/${id}`)
}

export async function createAgent(data: AgentCreate): Promise<Agent> {
  return apiClient<Agent>('/api/v1/agents', {
    method: 'POST',
    body: JSON.stringify(data),
  })
}

export async function updateAgent(id: string, data: AgentUpdate): Promise<Agent> {
  return apiClient<Agent>(`/api/v1/agents/${id}`, {
    method: 'PATCH',
    body: JSON.stringify(data),
  })
}

export async function deleteAgent(id: string): Promise<void> {
  await apiClient<void>(`/api/v1/agents/${id}`, {
    method: 'DELETE',
  })
}

export async function executeAgent(
  id: string,
  input: AgentExecuteInput
): Promise<AgentExecuteOutput> {
  return apiClient<AgentExecuteOutput>(`/api/v1/agents/${id}/execute`, {
    method: 'POST',
    body: JSON.stringify(input),
  })
}

export async function* executeAgentStream(
  id: string,
  input: AgentExecuteInput
): AsyncGenerator<string, void, unknown> {
  const response = await fetch(`${API_URL}/api/v1/agents/${id}/execute/stream`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
    },
    body: JSON.stringify(input),
  })

  if (!response.ok) {
    throw new Error(`Failed to execute agent: ${response.statusText}`)
  }

  const reader = response.body?.getReader()
  if (!reader) {
    throw new Error('Response body is not readable')
  }

  const decoder = new TextDecoder()
  let buffer = ''

  while (true) {
    const { done, value } = await reader.read()
    if (done) break

    buffer += decoder.decode(value, { stream: true })
    const lines = buffer.split('\n\n')
    buffer = lines.pop() || ''

    for (const line of lines) {
      if (line.startsWith('data: ')) {
        const data = line.slice(6)
        if (data === '[DONE]') {
          return
        }
        if (data.startsWith('[ERROR]')) {
          throw new Error(data.slice(8))
        }
        yield data
      }
    }
  }
}

export interface Execution {
  id: string
  agentId: string
  userInput: string
  result: string
  tokensUsed: number
  executionTimeMs: number
  createdAt: string
}

export async function getExecutions(
  agentId: string,
  skip = 0,
  limit = 20
): Promise<Execution[]> {
  return apiClient<Execution[]>(
    `/api/v1/agents/${agentId}/executions?skip=${skip}&limit=${limit}`
  )
}
